pkg/clients/edgeconnect/client.go

```diff
@@ -0,0 +1,278 @@
+	"golang.org/x/oauth2/clientcredentials"
+)
+
+type client struct {
+	ctx     context.Context
+	baseURL string
+	clientcredentials.Config
+}
+
+// Option can be passed to NewClient and customizes the created client instance.
+type Option func(client *client)
+
+// NewClient creates a REST client for the given API base URL
+// opts can be used to customize the created client, entries must not be nil.
+func NewClient(clientID, clientSecret string, options ...Option) (Client, error) {
+	c := &client{
+		Config: clientcredentials.Config{
+			ClientID:     clientID,
+			ClientSecret: clientSecret,
+			Scopes:       DefaultOauthScopes,
+			TokenURL:     DefaultTokenURL,
+		},
+	}
+
+	for _, opt := range options {
+		opt(c)
+	}
+
+	return c, nil
+}
+
+func NewClientWithProxy(clientID, clientSecret, proxy string) (Client, error) {
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	proxyUrl, err := url.Parse(proxy)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	transport.Proxy = func(req *http.Request) (*url.URL, error) {
+		return proxyUrl, nil
+	}
+	httpClient := http.Client{Transport: transport}
+	ctx := context.WithValue(context.Background(), oauth2.HTTPClient, httpClient)
+
+	return NewClient(clientID, clientSecret, WithContext(ctx))
+}
+
+func WithBaseURL(url string) func(*client) {
+	return func(c *client) {
+		c.baseURL = url
+	}
+}
+
+func WithTokenURL(url string) func(*client) {
+	return func(c *client) {
+		c.TokenURL = url
+	}
+}
+
+// WithContext can set context for client
+// NB: via context you can override default http client to add Proxy or CA certificates
+func WithContext(ctx context.Context) func(*client) {
+	return func(c *client) {
+		c.ctx = ctx
+	}
+}
+
+// ServerError represents an error returned from the server (e.g. authentication failure).
+type ServerError struct {
+	Code    int          `json:"code,omitempty"`
+	Message string       `json:"message,omitempty"`
+	Details DetailsError `json:"details"`
+}
+
+// DetailsError represents details of errors
+type DetailsError struct {
+	ConstraintViolations ConstraintViolations `json:"constraintViolations"`
+	MissingScopes        []string             `json:"missingScopes,omitempty"`
+}
+
+// ConstraintViolations represents constraint violation errors
+type ConstraintViolations struct {
+	Message           string `json:"message"`
+	Path              string `json:"path,omitempty"`
+	ParameterLocation string `json:"parameterLocation,omitempty"`
+}
+
+// Error formats the server error code and message.
+func (e ServerError) Error() string {
+	if len(e.Message) == 0 && e.Code == 0 {
+		return "unknown server error"
+	}
+
+	return fmt.Sprintf("edgeconnect server error %d: %s", int64(e.Code), e.Message)
+}
+
+type serverErrorResponse struct {
+	ErrorMessage ServerError `json:"error"`
+}
+
+func CloseBodyAfterRequest(response *http.Response) {
+	if response != nil && response.Body != nil {
+		err := response.Body.Close()
+		if err != nil {
```

Comment:
This error check here is either needless or error handling is missing. If nothing is done with the error, the function would return anyway  at this point.